### DISCUSSION TITLE
Hydration warnings for client-only data

### DISCUSSION THREAD

**Discussion by viridia** [solidjs/solid#1289] [Q&A]
I talked about this a bit on the Discord, but I wanted to write up a more lucid explanation. I've coded up a small example on stackblitz which demonstrates the problem I am seeing: https://stackblitz.com/edit/solid-start-hydration?file=src%2Froot.tsx

The use case here is a test framework for UI widgets, like storybook or react cosmos. There's a component under test, which is displayed in a preview pane. There is also a "tweak parameters" drawer which lets you fool around with the attributes of the component.

The component under test has access to a hooks-like API which lets it declare parameters, for example:

```tsx
export default () => {
  const params = createFixtureParams({
    min: { type: 'integer', minVal: -100, maxVal: 1000, default: 0 },
    max: { type: 'integer', minVal: -100, maxVal: 1000, default: 100 },
    step: { type: 'integer', minVal: 0, maxVal: 100, default: 0 },
  });
  return (
    <Group>
      <Slider
        min={params.min()}
        max={params.max()}
        step={params.step()}
```

The `createFixtureParams` function is connected to a Solid Store. Basically all this is doing is adding reactive properties to the store - effectively its creating signals and storing them in a map. The only unusual thing here is that this is happening as a side effect of rendering - but it's not that unusual since components normally call `createSignal()` or create stores during execution.

Also note that, like signals, the creation of these properties has to be synchronous since the signal accessors are used in the body of the component.

The tweak panel, on the other hand, looks at the store and uses it to decide what kind of tweak controls to build.

Now, this works perfectly well on the client. There's a bit of an extra rendering cycle depending on whether the preview panel or the tweak panel gets rendered first, since the tweak panel will have to be rendered twice, but it's hardly visible.

On the server, however, this has problems - if the tweak panel is rendered first (before the properties are created by the preview) it will render a blank property sheet. That's what gets hydrated. The client, unfortunately, expects there to be properties rendered, and emits a bunch of warnings about DOM nodes missing. Worse, the client tries to skip over the DOM nodes it thinks should be present, so what you get is a bizarre view of the tweak panel with half the elements missing. This persists until the user does a client-side navigate which re-renders the tweak panel.

The solution should be very simple - just disable rendering the tweak panel on the server entirely, and let the client build from scratch SPA style. There may be a slight flash as the rendering lags a bit, but since the tweak panel is in a pull-out drawer, often times this won't be visible anyway.

Unfortunately, disabling server-rendering of the tweak panel is not as easy as it seems - I've tried a bunch of things (onMount, queueMicrotask, setTimeout) to delay rendering and it insists on trying to hydrate the tweak panel anyway. For example, if I set the timeout to 5ms I get hydration warnings, but if I set it to 10ms I don't. That seems like a weird race condition.

The more general question here is how to deal with components that render based on client-side data - for example, preference settings in LocalStorage, which may hold things like which tree nodes are expended or which pull-out drawers are open. The way I have tended to code sites in the past is that "important" prefs are stored on the server, but it's not worth making an API endpoint for little nuances like autocompletion history or which rows are selected in the list view.

**Comment by viridia** [comment]
In the 10 days since I first posted this, I have learned a lot more about this issue - unfortunately, what I have learned is that the situation is much more pervasive than I realized.

There are a lot of UI components that utilize "throw-away persistent state". Take for example, a tree view: as the user expands and collapses tree nodes, the client needs to remember which nodes are expanded. This information needs to persist across a page reload: it's irritating to have to re-expand the tree every time you visit the page. However, it's also relatively unimportant data - less important than user preferences, less important even than "don't show this dialog again." If the data gets lost, and the user has to redo the expansions occasionally, it's not the end of the world.

Examples of widgets that use this kind of disposable state:

* Tree views
* Drawers
* "Collapse" widgets
* Splitter panes

The states for these widgets should *not* be stored on the server, for two reasons: first, they are not important enough to warrant columns in the database. And secondly, they are specific to a browser instance: if the user is using multiple devices, or even multiple browsers, they expect the states to be different - expanding a drawer on an iPad should not cause the drawer to be expanded on their desktop machine.

For "important" states like dark or light mode, which affect the entire UI, or where there are a small number of variables, cookies are the solution of choice since they can be read on the server.

However, it is not always possible to use cookies due to their limited size (4096 bytes). For example, if you have a large number of tree nodes, then LocalStorage or SessionStorage is a better option - but it means that the server can't access the state.

There are also URL params, but they too are limited in size. Also, the behavior of the "back" button will affect the state of the widgets, which is probably not what you want.

Given the choice to use LocalStorage, it means that server-side rendering is not going to work properly. A tree widget rendered on the server will not know which nodes are to be expanded; when the client attempts to hydrate this tree, it's going to get lots of warnings since the DOMs are not the same.

For these kinds of cases, I'd like to be able to simply disable SSR for that part of the app. In other words, as a designer of the UI, I can decide that a flash of content - caused by rendering on the client after JS has been loaded - is acceptable as long as it is confined to components that are not the main focus of attention. As long as "most" of the page is being rendered with SSR, I can live with the notion that some sidebar is being rendered client-only. As a designer, I can even come up with ways to make this transition look attractive, possibly with some sort of animated placeholder.

So basically what is needed is a way to disable hydration selectively.

However, actually accomplishing this in code is trickier than it sounds. The standard way of "client-only" rendering is to have a flag which is set in an `onMount` or `createEffect` callback: only render the content if the flag is set, meaning that we're running on the client.

However, these methods break badly if the component being rendered has a `<Suspense>` tag, because (AFAICT) Suspense resumption re-starts the hydration process. It means that your client-only widget is still going to try and hydrate, which will fail.

There may be other failure modes caused by async logic, but I have not had the time to discover them.

My first point is that it is not easy or obvious for a user of Solid to know the correct formula for creating a client-only widget. And my second point is that this is not a rare or esoteric use case.

So my ask here, is for someone on the Solid team to put some thought into this, and come up with a robust pattern that developers like me can adopt.

Ideally, this would simply be a tag or directive that tells SSR "don't try and reconcile the server DOM and the client DOM - simply overwrite the DOM with what the client rendered." - and which will work even with Suspense.

**Comment by ryansolid** [comment]
Yeah it's a bit tricky. I need to look into this to figure out why `onMount` type approaches didn't work. My thinking is its more than DOM reconciliation we need to worry about if we are going to exit hydration mode. Ideally the boundaries of hydration are at the right points and we can just execute them there. It's a lot easier to do that from a root of execution than in the middle or we risk entanglement. I will probably need to spend some time with this one.
